prov/util/src/rxm_av.c

```diff
@@ -227,14 +242,27 @@ static int rxm_av_remove(struct fid_av *av_fid, fi_addr_t *fi_addr,
 		if (!av_entry)
 			continue;
 
+		if (av->util_av.remove_handler) {
+			peer = ofi_av_addr_context(&av->util_av, fi_addr[i]);
+			(*peer)->refcnt++; /* prevent peer from being freed */
+			ofi_mutex_lock(&av->util_av.ep_list_lock);
+			dlist_foreach(&av->util_av.ep_list, item) {
+				util_ep = container_of(item, struct util_ep,
+						       av_entry);
+				av->util_av.remove_handler(util_ep, *peer);
+			}
+			ofi_mutex_unlock(&av->util_av.ep_list_lock);
+			_util_put_peer(*peer);
+		}
+
 		if (!ofi_atomic_dec32(&av_entry->use_cnt)) {
 			rxm_put_peer_addr(av, fi_addr[i]);
 			HASH_DELETE(hh, av->util_av.hash, av_entry);
 			ofi_ibuf_free(av_entry);
 		}
 	}
-
 	ofi_mutex_unlock(&av->util_av.lock);
+
```

Comment:
The locking does not work.  This call is forming a lock order of AV lock -> AV ep_list_lock -> EP lock.  The current code has EP lock -> AV lock.  The purpose of the AV ep_list_lock is to support the lock ordering of AV ep_list_lock -> EP lock, because we can't hold the AV lock before acquiring the EP lock.

Data transfers start with an EP.  So, they acquire the EP lock first.  From there, the transfer needs to get information about the peer.  That data come from the AV, which uses the AV lock.  This creates the EP lock -> AV lock ordering.

You need to drop the AV lock prior to invoking the remove_handler().